vncpasswd add pwquality complexity rule check

prownd opened the pull request:
Add a password complexity strategy for VNC user authentication
enhance password quality verification checks, and enhance security

Additionally, Chinese translation has been added

CendioOssman commented:
Thank you for your contribution!

I'm concerned how well a pwquality will work for VNC passwords, given that they are limited to 8 characters. #370 would probably need to be implemented first. Otherwise the user might come up with a long and secure to keep pwquality happy, but overlook the fact that the actual used password is much shorter.

It should probably also just be a warning when pwquality fails, not an error, given that it is difficult to get a good password with so few characters.

prownd commented:
Yes, due to the VNC protocol, the actual length of the VNC password used is 8 characters.
The first version I modified did indeed limit the maximum length to 8 characters. However, considering that there has always been no limit on the maximum length when setting passwords,  so I removed the limit of 8 characters.
According to your suggestion, when pwquality fails, there will only be an alarm prompt. Instead of terminating password settings?

prownd commented:
> Thank you for your contribution!
> 
> I'm concerned how well a pwquality will work for VNC passwords, given that they are limited to 8 characters. #370 would probably need to be implemented first. Otherwise the user might come up with a long and secure to keep pwquality happy, but overlook the fact that the actual used password is much shorter.
> 
> It should probably also just be a warning when pwquality fails, not an error, given that it is difficult to get a good password with so few characters.

According to your opinion, when setting a password, add an 8-character limit.
Can I modify and submit it again ?

prownd commented:
1.  Use libpwquality to check password quality and complexity, improving security.
    The validation rules for libpwquality are as follows:
    - The minimum password length is 6 and the maximum is 8
    -  Password characters repeat once
    -  At least 3 types of characters are required
2.  Add i18n internationalization support and Chinese localization translation.

prownd commented:
> 1. Use libpwquality to check password quality and complexity, improving security.
>    The validation rules for libpwquality are as follows:
>    
>    * The minimum password length is 6 and the maximum is 8
>    * Password characters repeat once
>    * At least 3 types of characters are required
> 2. Add i18n internationalization support and Chinese localization translation.

@CendioOssman Could you please take a look and see if there is anything else that needs to be added.

CendioOssman commented on po/zh_CN.po:
```diff
@@ -703,3 +703,31 @@ msgstr "正在监听端口 %d"
+
+#: unix/vncpasswd/vncpasswd.cxx:181
+msgid "Passwords don't match - try again\n"
+msgstr "密码不匹配,请重试\n"
```
I'm afraid our translations are handled by the translation project. As a result, we cannot modify these files directly or they'll get overwritten. Please send your translations to them once a beta is out of TigerVNC with your changes in it.

prownd replied:
Okay, I understand. Maybe I didn't consider it before.

CendioOssman commented on unix/vncpasswd/CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@ add_executable(vncpasswd
 
 target_include_directories(vncpasswd PUBLIC ${CMAKE_SOURCE_DIR}/common)
-target_link_libraries(vncpasswd tx rfb os)
+target_link_libraries(vncpasswd tx rfb os pwquality)
```
How commont is pwquality? Is it available for all distributions? Otherwise it will need to be an optional dependency.

Could you please also add it to BUILDING.txt?

prownd replied:
I checked and tested the pwqaulity.pc of the rpm and deb based distribution (centos/debian)
The compilation link parameters are all "- l pwquality"
It should be universal, and I will test and check other distributions later.

prownd replied:
Compiling on Debian requires installing the libpwquality-dev package.

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -37,6 +37,15 @@
 
+#include <pwquality.h>
+#include <locale.h>
+#include <libintl.h>
```
Please split adding translation support as a separate commit. It makes it easier to see what parts are associated.

prownd replied:
Okay, it really should be separated.

prownd replied:
sorry，
Is it adding a new commit in the current PR. 
Or create a new PR to handle translation.

CendioOssman replied:
It can be part of this PR as long as it's a separate commit.

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -37,6 +37,15 @@
 
+#include <pwquality.h>
+#include <locale.h>
+#include <libintl.h>
```
Let's try to use the same convenience wrappers that we have for vncviewer. Please refer to the files `i18n.h` and `gettext.h` there.

prownd replied:
Okay, I'll try to modify it.

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -37,6 +37,15 @@
+
+#define PACKAGE_NAME "tigervnc"
+#define PACKAGE_NAME2 "libpwquality"
+#define LOCALE_DIR  "/usr/share/locale"
```
This should come from CMake. Please refer to vncviewer on how to get this.

prownd replied:
I was not very familiar with CMake before
Okay, I will make the necessary modifications according to your suggestions

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -99,6 +108,39 @@ static int encrypt_pipe() {
+	pwquality_set_int_value(pwq, PWQ_SETTING_MIN_LENGTH, 6);
+	pwquality_set_int_value(pwq, PWQ_SETTING_MAX_SEQUENCE, 8);
+	pwquality_set_int_value(pwq, PWQ_SETTING_MAX_REPEAT, 1);
+	pwquality_set_int_value(pwq, PWQ_SETTING_MIN_CLASS, 3);
```
This doesn't seem proper to override. Isn't this configured by the system administrator in `/etc/security/pwquality.conf`?

prownd replied:
The default rules are in pwquality.conf
If specified in the code, the rules will be rewritten.

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -109,10 +151,22 @@ static std::vector<uint8_t> readpassword() {
         exit(1);
       }
-      fprintf(stderr,"Password must be at least 6 characters - try again\n");
+      fprintf(stderr,_("Password must be at least 6 characters - try again\n"));
```
This is now policy that could conflict with what pwquality wants. It should be removed if we always use pwquality, or used as a fallback if pwquality ends up being optional.

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -109,10 +151,22 @@ static std::vector<uint8_t> readpassword() {
+    }
+
+    if (first.size() > 8) {
+      fprintf(stderr,_("Password should not be greater than 8 characters\nBecause only 8 valid characters are used - try again\n"));
```
It's best to avoid formatting in the strings as it makes it difficult to translate.

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -109,10 +151,22 @@ static std::vector<uint8_t> readpassword() {
+      continue;
+    }
+
+    if (first.size() > 8) {
```
Nice! But could you split this to a separate commit? It's independent of using pwquality.

prownd replied:
Perhaps they should be submitted separately

CendioOssman replied:
A separate commit, yes. But you don't need to open a separate PR for such a small fix.

CendioOssman commented on unix/vncpasswd/vncpasswd.cxx:
```diff
@@ -133,6 +187,11 @@ static std::vector<uint8_t> readpassword() {
 {
+  setlocale(LC_ALL, "");
+  bindtextdomain(PACKAGE_NAME, LOCALE_DIR);
+  bind_textdomain_codeset(PACKAGE_NAME2, "UTF-8");
```
This seems wrong. Why is this needed? We should not need to mess with a library's translation handling. And UTF-8 might not be the correct character set for the terminal.

CendioOssman reviewed (changes requested):
Looks really good! There are some tweaks that are needed, though. Please have a look at my comments.

prownd commented:
@CendioOssman  
I have made modifications according to your suggestions. 
Could you please review it to see if it is correct.

prownd commented:
Could you please take a moment to take a  review  again .  @CendioOssman

prownd commented:
@CendioOssman Please review again.
Is your suggestion to use the default/etc/security/pwquality.conf configuration rules?
But as mentioned earlier, the actual maximum length of a password is only 8, which may have generated a very long and complex password, but only 8 lengths were actually used.
So it is very necessary to specify some rules in the code.

Looking forward to your reply and suggestions.

CendioOssman commented:
Sorry for the late response. I've been away for a few weeks.

There are still some open requests that haven't been addressed. Please have a look at them and see if we can get the last bits fixed.

CendioOssman commented:
> Is your suggestion to use the default/etc/security/pwquality.conf configuration rules?

Yes, that is correct.

> But as mentioned earlier, the actual maximum length of a password is only 8, which may have generated a very long and complex password, but only 8 lengths were actually used.
> So it is very necessary to specify some rules in the code.

If the system policy doesn't allow passwords that are compatible with VNC, then the system should refuse to continue. It shouldn't silently change the policy and allow things that the user thinks have been blocked.

As for only 8 characters being used, your previous suggestion had a check for that. But it seems to have been dropped?

prownd commented:
> @CendioOssman Please review again. Is your suggestion to use the default/etc/security/pwquality.conf configuration rules? But as mentioned earlier, the actual maximum length of a password is only 8, which may have generated a very long and complex password, but only 8 lengths were actually used. So it is very necessary to specify some rules in the code.
> 
> Looking forward to your reply and suggestions.



> > Is your suggestion to use the default/etc/security/pwquality.conf configuration rules?
> 
> Yes, that is correct.
> 
> > But as mentioned earlier, the actual maximum length of a password is only 8, which may have generated a very long and complex password, but only 8 lengths were actually used.
> > So it is very necessary to specify some rules in the code.
> 
> If the system policy doesn't allow passwords that are compatible with VNC, then the system should refuse to continue. It shouldn't silently change the policy and allow things that the user thinks have been blocked.
> 
> As for only 8 characters being used, your previous suggestion had a check for that. But it seems to have been dropped?



> > Is your suggestion to use the default/etc/security/pwquality.conf configuration rules?
> 
> Yes, that is correct.
Okay, then use the default pwquality-oriented. conf rule
> 
> > But as mentioned earlier, the actual maximum length of a password is only 8, which may have generated a very long and complex password, but only 8 lengths were actually used.
> > So it is very necessary to specify some rules in the code.
> 
> If the system policy doesn't allow passwords that are compatible with VNC, then the system should refuse to continue. It shouldn't silently change the policy and allow things that the user thinks have been blocked.
> 
> As for only 8 characters being used, your previous suggestion had a check for that. But it seems to have been dropped?

Because it's independent of using pwquality
Your suggestion is to submit a separate commit


> > Is your suggestion to use the default/etc/security/pwquality.conf configuration rules?
> 
> Yes, that is correct.
> 
> > But as mentioned earlier, the actual maximum length of a password is only 8, which may have generated a very long and complex password, but only 8 lengths were actually used.
> > So it is very necessary to specify some rules in the code.
> 
> If the system policy doesn't allow passwords that are compatible with VNC, then the system should refuse to continue. It shouldn't silently change the policy and allow things that the user thinks have been blocked.
> 
> As for only 8 characters being used, your previous suggestion had a check for that. But it seems to have been dropped?

prownd commented:
> > Is your suggestion to use the default/etc/security/pwquality.conf configuration rules?
> 
> Yes, that is correct.

Okay, then use the default pwquality.conf rule

>
> > But as mentioned earlier, the actual maximum length of a password is only 8, which may have generated a very long and complex password, but only 8 lengths were actually used.
> > So it is very necessary to specify some rules in the code.
> 
> If the system policy doesn't allow passwords that are compatible with VNC, then the system should refuse to continue. It shouldn't silently change the policy and allow things that the user thinks have been blocked.
> 
> As for only 8 characters being used, your previous suggestion had a check for that. But it seems to have been dropped?

Because it's independent of using pwquality
Your suggestion is to submit a separate commit

prownd commented:
@CendioOssman  According to your suggestion, it has been modified. 
Could you please review it again

CendioOssman reviewed (approved):
Looks good! Thank you for getting this done.
